src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -319,22 +319,44 @@ namespace ts.codefix {
+            //
+            // To:
+            //   let onFulfilledArg;
+            //   try {
+            //     onFulfilledArg = await x();
+            //   } catch (onRejectedArg) {
+            //     /* on rejected */
+            //     return;
+            //   }
+            //   /* on fulfilled */
             const onRejectedArgumentName = getArgBindingName(onRejected, transformer);
-            const tryBlock = createBlock(transformExpression(node.expression, transformer, onFulfilledArgumentName).concat(transformationBody));
-            const transformationBody2 = getTransformationBody(onRejected, prevArgName, onRejectedArgumentName, node, transformer);
+            const onFulfilledArgType = onFulfilledArgumentName && !transformer.isInJSFile && transformer.checker.getTypeAtLocation(getNode(onFulfilledArgumentName));
+            const onFulfilledArgTypeNode = onFulfilledArgType && !(onFulfilledArgType.flags & TypeFlags.Any) ? transformer.checker.typeToTypeNode(onFulfilledArgType) : undefined;
+            const onFulfilledArgVariableDeclaration = createVariableOrAssignmentOrExpressionStatement(onFulfilledArgumentName, /*rightHandSide*/ undefined, onFulfilledArgTypeNode);
+            const onFulfilledBody = getTransformationBody(onFulfilled, prevArgName, onFulfilledArgumentName, node, transformer);
+            const tryBlock = createBlock(transformExpression(node.expression, transformer, onFulfilledArgumentName));
+            const onRejectedBody = ensureReturnOrThrow(getTransformationBody(onRejected, prevArgName, onRejectedArgumentName, node, transformer));
             const catchArg = onRejectedArgumentName ? isSynthIdentifier(onRejectedArgumentName) ? onRejectedArgumentName.identifier.text : onRejectedArgumentName.bindingPattern : "e";
             const catchVariableDeclaration = createVariableDeclaration(catchArg);
-            const catchClause = createCatchClause(catchVariableDeclaration, createBlock(transformationBody2));
-
-            return [createTry(tryBlock, catchClause, /* finallyBlock */ undefined)];
+            const catchClause = createCatchClause(catchVariableDeclaration, createBlock(onRejectedBody));
+            return onFulfilledArgVariableDeclaration.concat(createTry(tryBlock, catchClause, /* finallyBlock */ undefined), onFulfilledBody);
         }
 
+        const transformationBody = getTransformationBody(onFulfilled, prevArgName, onFulfilledArgumentName, node, transformer);
         return transformExpression(node.expression, transformer, onFulfilledArgumentName).concat(transformationBody);
     }
 
+    function ensureReturnOrThrow(statements: readonly Statement[]): readonly Statement[] {
+        const lastStatement = lastOrUndefined(statements);
+        if (lastStatement?.kind !== SyntaxKind.ReturnStatement && lastStatement?.kind !== SyntaxKind.ThrowStatement) {
+            return statements.concat(createReturn());

```

Comment:
Looks like this could be unreachable (e.g. if the last statement is actually an if-else in which both branches return)?

Reply:
Hm, you’re right. I’m not sure there’s an efficient way to know what to do here.